### PR TITLE
Feature/tech 116 dagster deploy step

### PR DESCRIPTION
**Pull request by kgrunert**
- [ ] run `mpyl project lint` for all projects and correct the schema changes from `kubernetes.secrets` -> `dagster.secrets`
- [ ] enable back linting in Jenkinsfile

----
### 📕 [TECH-116](https://vandebron.atlassian.net/browse/TECH-116) Implement Dagster Deploy step katringrunert@vandebron.nl 
Convert `Dagster Deploy` step in mpl-modules to mpyl.

Modules to convert:

* DagsterDeploy
* DagsterUtil

**T-Shirt size: M**

🏗️ Build [16](https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline%20-%20Test/job/PR-146/16/display/redirect) ✅ Successful, started by _Katrin Grunert_  
🚀 *[cloudfront-service](https://cloudfront-service-146.test.nl/)*, *example-dagster-user-code*, *job*, *kong-sync*, *[nodeservice](https://nodeservice-146.test.nl/)*, *[sbtservice](https://sbtservice-146.test.nl/)*, *sparkJob*  

[TECH-116]: https://vandebron.atlassian.net/browse/TECH-116?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (kgrunert) [inline src/mpyl/steps/deploy/k8s/__init__.py:49]: use logger

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [inline src/mpyl/steps/deploy/dagster.py:69]: it is safe to assume that DagsterDeploy  is executed for one project

can be found in `step_input.project`

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [inline src/mpyl/steps/deploy/dagster.py:51]: export it into something like `to_spark_body()`

Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py:51]: hard coded names of the registry -> use config

Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py]: didnt find any field for `imagePullSecrets` tho. ok to keep it hard coded?

File: .github/workflows/build-package.yml
Comment (Jorg88) [inline .github/workflows/build-package.yml]: is this intentional?

File: .github/workflows/build-package.yml
Comment (koeves) [inline .github/workflows/build-package.yml:1]: These are just formatting changes, maybe you have different rules then we do. You can configure yaml formatted to not indent lists like that in pycharm

File: src/mpyl/project.py
Comment (koeves) [inline src/mpyl/project.py:273]: What kind of secrets are these? Can you link a project yaml example? Because we have now sealed secrets, but also "normal" k8s secrets (done in #165), and now this is a third one?

File: src/mpyl/project.py
Comment (koeves) [inline src/mpyl/project.py:318]: Is repo mandatory? If so then I'd say do
```suggestion
    @staticmethod
    def from_config(values: dict):
        repo = values.get("repo")
        if not repo:
          raise KeyError("Dagster config needs to have repo field set")
        return Dagster(
            repo=repo
        )
```

File: src/mpyl/project.py
Comment (koeves) [inline src/mpyl/project.py:419]: dagster configuration instead of kubernetes configuration?

File: src/mpyl/steps/deploy/dagster.py
Comment (koeves) [inline src/mpyl/steps/deploy/dagster.py:37]: This method could be in a helm utilities module

**Review by Jorg88 — COMMENTED** [review]
Not sure if you spoke about this with Daniel, but we might want to add some tests for the new functionalities you added and maybe even a test project?

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (koeves) [inline src/mpyl/steps/deploy/k8s/__init__.py:56]: Shouldn't we signal failure here so we can fail the execute step?

**Review by koeves — CHANGES_REQUESTED** [review]
Nice work! Some smaller comments

**Comment by koeves** [comment]
> Not sure if you spoke about this with Daniel, but we might want to add some tests for the new functionalities you added and maybe even a test project?

Yes I forgot to mention this yesterday, we should try to add a small example Dagster project to https://github.com/Vandebron/mpyl/tree/99e36e3e5813d4aa5178da1158c59fea79649804/tests/projects and add a roundtrip test for the values.yaml like in test_k8s.py. Aand some smaller unit test for the new methods where feasible would be nice!

File: .github/workflows/build-package.yml
Comment (Jorg88) [reply .github/workflows/build-package.yml:1]: I also commented on this but the changes are gone already 😉

File: src/mpyl/project.py
Comment (kgrunert) [reply src/mpyl/project.py:273]: https://github.com/Vandebron/scripting/blob/7256aa60b4afef60dadf227b323ef2c55a19a4db/python/production_correction/deployment/project.yml

This is how it looks for example

Comment (koeves) [reply src/mpyl/project.py:273]: Oof that's indeed a third one.. @Jorg88

Comment (kgrunert) [reply src/mpyl/project.py:273]: They get reference in the user-code-deployment of dagster.

I would love to make it fit into the strcutre of #165

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/__init__.py:56]: yes indeed 👍

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py]: done in
[34b4331](https://github.com/Vandebron/mpyl/pull/146/commits/34b4331b34b4fa1ada15dd88ce7115eb7e08464c)

File: src/mpyl/project.py
Comment (kgrunert) [reply src/mpyl/project.py]: done in
[34b4331](https://github.com/Vandebron/mpyl/pull/146/commits/34b4331b34b4fa1ada15dd88ce7115eb7e08464c)

File: src/mpyl/project.py
Comment (kgrunert) [reply src/mpyl/project.py]: done in
[34b4331](https://github.com/Vandebron/mpyl/pull/146/commits/34b4331b34b4fa1ada15dd88ce7115eb7e08464c)

File: src/mpyl/project.py
Comment (kgrunert) [reply src/mpyl/project.py:273]: those secrets get set as `envSecret` in the user-code-deployments (see https://github.com/dagster-io/dagster/blob/3d0884e617e58bc1e97f9f40c4096e7189bfb678/helm/dagster/values.yaml#L340)

Comment (koeves) [reply src/mpyl/project.py:273]: Ah indeed I see from git blame, this secret's mechanism is added by Pieter for deploying Dagster. In that case, we should try to combine it with the approach of #165, or move it to the dagster config if it's possible

Comment (kgrunert) [reply src/mpyl/project.py:273]: moving them to dagster seems feasible,
i think the future idea was to marry sealed-secrets and dagster

Comment (koeves) [reply src/mpyl/project.py:273]: That'd be the best!

File: src/mpyl/steps/deploy/dagster.py
Comment (koeves) [reply src/mpyl/steps/deploy/dagster.py]: I meant like use the registry name from the mpyl_config.yaml

Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py]: oh yes, i also found that and applied it, but then i noticed that the `imagePullSecret` will probably stay hardcoded

**Comment by kgrunert** [comment]
> Yes I forgot to mention this yesterday, we should try to add a small example Dagster project to https://github.com/Vandebron/mpyl/tree/99e36e3e5813d4aa5178da1158c59fea79649804/tests/projects and add a roundtrip test for the values.yaml like in test_k8s.py. Aand some smaller unit test for the new methods where feasible would be nice!

`test_dagster.py` and `test_helm_utilities.py` give a small roundtrip from project.yml -> user-code-values.yaml and its used utility methods

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/__init__.py:60]: like discussed i am returning `Output`s in these function that can get evaluated by the DagsterDeploy step (and exiting early if failure)

File: .github/workflows/build-package.yml
Comment (kgrunert) [inline .github/workflows/build-package.yml:165]: revert change

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [inline src/mpyl/steps/deploy/dagster.py:131]: use `output` consistently

File: src/mpyl/steps/deploy/k8s/resources/dagster.py
Comment (Jorg88) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: We already have similar logic in `to_service_account` btw, `image_pull_secrets=[V1LocalObjectReference("bigdataregistry")]`

File: src/mpyl/steps/deploy/dagster.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/dagster.py:56]: You can use `f"-{step_input.run_properties.versioning.identifier}"`

File: src/mpyl/steps/deploy/dagster.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/dagster.py:109]: `replaced_config_map_result` ?

File: src/mpyl/steps/deploy/dagster.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/dagster.py:124]: Why are we doing this another time if the first try succeeded?

File: src/mpyl/steps/deploy/dagster.py
Comment (Jorg88) [reply src/mpyl/steps/deploy/dagster.py:131]: Hmm, the benefit of this approach instead of just having a try catch block around this entire `if deploy_result.success and not step_input.dry_run` code block is that you have more precise error messages on failures. But it does make the code a bit messy, not sure what's a better approach actually.

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/k8s/__init__.py:51]: This success message will never be used right, since we only return on failure and otherwise we always end up returning `deploy_result` (in `DeployDagster.execute`)? Same for the `replace_config_map` function below

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/k8s/__init__.py:57]:
```suggestion
            message=f"Exception when calling AppsV1Api -> patch_namespaced_deployment: {api_exception}\n"
```

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/k8s/__init__.py:94]: I think I would pass the `api`, reduces code duplication (with the functions below) and ensures the config/api is only loaded/instantiated once

File: src/mpyl/steps/deploy/k8s/helm.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/k8s/helm.py:31]:
```suggestion
def repo_update(logger: Logger):
```

File: src/mpyl/steps/deploy/k8s/helm.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/k8s/helm.py:98]: We could combine this part with the `install` function. Create a base function that has all the duplicated logic and pass the parts of the cmd that are different to append to the base cmd.

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py:40]: at the moment the hard-coded namespace is the best. as dagster deploys can only run in the namespace where the dagster-instance is also running.

i created a follow-up card that tackles this issue:  https://vandebron.atlassian.net/browse/TECH-519

File: src/mpyl/steps/deploy/k8s/resources/dagster.py
Comment (koeves) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: I'd refactor that one as well then

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py:124]: we are doing restarts of two different instances (`dagit` and `daemon`), but if one restart is already failing i would like to have this result to be the error message that indicates the failure of the build

Comment (Jorg88) [reply src/mpyl/steps/deploy/dagster.py:124]: in that case.. did you mean to pass a different string on the next line?

Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py:124]: WOOOOOOOW 😂 🤦  sharp eye! thanks :D

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/__init__.py:51]: hmmm youre absolutely right. What do you think is the better approach. Moving the logging of the success into the `DagsterDeploy` step or flattening all success-Outputs into one Output that `DagsterDeploy` is returning?

Comment (koeves) [reply src/mpyl/steps/deploy/k8s/__init__.py:51]: Agreed, let's put some logging and flatten the outputs

File: src/mpyl/steps/deploy/k8s/helm.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/helm.py]: refactored into `update_repo`

File: src/mpyl/steps/deploy/k8s/resources/dagster.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: i took  a look into `to_service_account` , but there is no parallel that is refactorable.  in `to_service_account` we just make use of the classes defined in the python kubernetes package (`V1ServiceAccount` , etc...).
the imagePullSecretName "bigdataregistry" is also hardcoded there.
So i guess i can add an extra field in the `dockerconfig` a field for the pullSecret. and then we can remove the hardcoded name?

Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: or am i overlooking something?

Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: Took a look and it seems like we got that name already in the DockerConfig/Registry, so will give that a try

Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: so, in that sense it wouldnt be smart user docker_config.user_name?

Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: and also i have NO idea when those secrets are being created. it seems every namespace has them

Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: sorry for the train of thought in here hahaha 🚄 💨

Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:23]: Checked in with Martijn from platform. Those secrets are being created by rancher (or we configured rancher to do exactly that). So the name wont change anytime soon! Then for ease, i would keep that hardcoded actually 😄

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py]: would `config_map_update_result` be better or what was your question ?

File: src/mpyl/steps/deploy/k8s/__init__.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/__init__.py]: done in 
65d5d77

Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/__init__.py]: and logging taken care of in 
2449da9

File: src/mpyl/steps/deploy/k8s/helm.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/helm.py]: done in 
34abdda

File: src/mpyl/steps/deploy/k8s/helm.py
Comment (SamTheisens) [inline src/mpyl/steps/deploy/k8s/helm.py:107]: I'm wondering 🤔 if you could use `write_chart` perhaps? For our standard HELM chart writing logic we already create a `values.yml` file in the usual place, albeit empty.
Perhaps it would be more intuitive to add the possibility to really write values there?

File: src/mpyl/steps/deploy/k8s/resources/dagster.py
Comment (SamTheisens) [inline src/mpyl/steps/deploy/k8s/resources/dagster.py:21]: Are any of these values specific to our setup?

File: src/mpyl/steps/deploy/k8s/resources/dagster.py
Comment (SamTheisens) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:38]: I think we should move this hardcoded, and Vandbebron specific, secret name to `mpyl_config.yml`. In the `kubernetes` section perhaps?

**Review by SamTheisens — CHANGES_REQUESTED** [review]
Looks good in general! Mostly nitpicky comments.
I'm not a fan of the stateful finicky nature of dagster deployments. But apparently that's the state of the art.
Somebody here has attempted an ArgoCD based solution https://github.com/dagster-io/dagster/issues/3099, perhaps that could give us some inspiration?

**Comment by SamTheisens** [comment]
> Looks good. Just keep in mind that you need to migrate the project yamls of the dagster projects so the secrets are at the right place

Perhaps we should build a migration command into mpyl that migrates `project.yml` to the latest version. In that case we'd need a version field in `project.yml` too.

File: src/mpyl/steps/deploy/k8s/resources/dagster.py
Comment (SamTheisens) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:38]: I moved made the pull secrets configurable on this PR https://github.com/Vandebron/mpyl/pull/185/commits/ebe774bf698f7caa463bd168368d2b9cbfac68ea

File: src/mpyl/steps/deploy/dagster.py
Comment (Jorg88) [reply src/mpyl/steps/deploy/dagster.py]: my point was that you use `config_map` everywhere else and here `configmap` :)

File: src/mpyl/steps/deploy/dagster.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/dagster.py:172]: Shouldn't you remove the other returns (lines 170, 165, 146) and add them to `dagster_deploy_results` instead, in order to fully utilize `__flatten_result_messages`?

File: src/mpyl/steps/deploy/k8s/helm.py
Comment (Jorg88) [inline src/mpyl/steps/deploy/k8s/helm.py:95]: These logs actually seem redundant since `custom_check_output` already does `logger.info(f"Executing: '{command_argument}'")`

Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/helm.py:95]: you are right!

File: src/mpyl/utilities/helm/__init__.py
Comment (kgrunert) [reply src/mpyl/utilities/helm/__init__.py:8]: nice one,  i would export the dagster-specific shortening of the release name into a separate function then

File: src/mpyl/steps/deploy/k8s/resources/dagster.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:21]: i think so. (except HELM_CHART_REPO and CHART_NAME) -> as a follow up i would like to also specific the `DagsterDeploy`-step as a `DagsterUserCodeDeploy` step, but for the sake of our transition from MPL to MPyL, we have this intermediate state.

Do you think mpyl-config, or project.yml would be a place for the other constants?

**Comment by SamTheisens** [comment]
@kgrunert I tried to rebase. That was pretty hopeless. A merge wasn't trivial either. Perhaps it's quickest and cleanest to create a new branch and re-apply the changes?

File: src/mpyl/steps/deploy/k8s/resources/dagster.py
Comment (koeves) [reply src/mpyl/steps/deploy/k8s/resources/dagster.py:21]: a dagster section in the mpyl config seems like a good choice

**Comment by kgrunert** [comment]
> @kgrunert I tried to rebase. That was pretty hopeless. A merge wasn't trivial either. Perhaps it's quickest and cleanest to create a new branch and re-apply the changes?

Rebasing was a nightmare, but the Github's "Resolve Conflicts" gave a really helpful overview on how to merge.

File: src/mpyl/steps/deploy/k8s/helm.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/k8s/helm.py]: i have extending the `write_chart`-argument list by a dict of `values`, for the normal helm-chart-writes we pass an empty dict, and for the dagster-helm-chart, we are passing  an empty chart and the values that need to be written to file

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py:183]: [d88a106](https://github.com/Vandebron/mpyl/pull/146/commits/d88a1066332f567615ea24bfb55f17842cf79cfa)

addressed this here

Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py:183]: looks way nicer with processing your comment!

File: src/mpyl/steps/deploy/dagster.py
Comment (kgrunert) [reply src/mpyl/steps/deploy/dagster.py]: yeah i do, i already had a flattening of outputs in place but added them to that structure too. so we get an early exit

File: src/mpyl/steps/deploy/dagster.py
Comment (Jorg88) [reply src/mpyl/steps/deploy/dagster.py]: (so basically just rename the variable 😛 )

**Review by SamTheisens — COMMENTED** [review]
Looks great 👍 . A few small comments

File: src/mpyl/schema/mpyl_config.schema.yml
Comment (SamTheisens) [reply src/mpyl/schema/mpyl_config.schema.yml]: Ah, it looks like you unified them in a later commit. Never mind.

File: src/mpyl/schema/mpyl_config.schema.yml
Comment (kgrunert) [reply src/mpyl/schema/mpyl_config.schema.yml]: outdated

File: src/mpyl/schema/mpyl_config.schema.yml
Comment (kgrunert) [reply src/mpyl/schema/mpyl_config.schema.yml]: unlucky naming for `ImagePullSecrets` but not relevant right now

**Comment by kgrunert** [comment]
> Looks great 👍 . A few small comments

Processed your comments 👍 Just need your approval for the merge now 😄

**Comment by kgrunert** [comment]
> Looks great 👍 . A few small comments

Addressed your last comments 👍  once you approve we can merge 🎉

**Review by Jorg88 — APPROVED** [review]

**Review by SamTheisens — APPROVED** [review]
I think this PR is worth a mention in the release notes. But that's also something you can add after merge.

**Pull request merged**
